enarx-pr-assign

```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/python3
+# SPDX-License-Identifier: Apache-2.0
+
+from github import Github
+import os
+
+RESPONSIBILITIES = {
+    'CHANGES_REQUESTED': lambda author, reviewer, requests: reviewer if reviewer in requests else author,
+    'DISMISSED': lambda author, reviewer, requests: reviewer,
+    'APPROVED': lambda author, reviewer, requests: None,
+    'COMMENTED': lambda author, reviewer, requests: reviewer,
+    'PENDING': lambda author, reviewer, requests: reviewer,
+}
+
+github = Github(os.environ['GITHUB_TOKEN'])
+
+# Return a dictionary with the latest review for every reviewer on a given PR.
+def get_latest_reviews(pr):
+    return {r.user.id: r for r in pr.get_reviews()}
+
+def get_requested_reviewers(pr):
+    # Collect sets of the latest reviews.
+    latest_reviews = get_latest_reviews(pr)
+    changes_requested_reviews = {u.user.id: u for (k, u) in latest_reviews.items() if u.state == "CHANGES_REQUESTED"}
+    
+    # Yield reviewers whose Changes Requested reviews have been addressed.
+    for e in pr.get_issue_events():
+        if(e.requested_reviewer is not None and
+        e.requested_reviewer.id in changes_requested_reviews and
+        e.created_at > changes_requested_reviews[e.requested_reviewer.id].submitted_at):
+            # If the above condition is true, the PR author must have requested
+            # a re-review after the latest submitted CHANGES_REQUESTED review
+            # from an individual reviewer, implying that their review has been
+            # addressed and the reviewer needs to re-review.
+            yield e.requested_reviewer.login
+
+    # Yield reviewers who have been requested, but have not left any kind of
+    # review yet.
+    for reviewer in pr.get_review_requests()[0]:
+        if reviewer.id not in latest_reviews.keys():
+            yield reviewer.login
+
+def handle(issue, pr, responsible):
+    assignees = {a.login for a in pr.assignees}
+
+    remove = assignees - responsible
+    needed = responsible - assignees
+
+    repo = issue.repository
+    print(f"{repo.organization.login}/{repo.name}#{pr.number}:", end='')
+    for u in needed:
+        print(f" +{u}", end='')
+    for u in remove:
+        print(f" -{u}", end='')
+    print()
+
+    if remove:
+        pr.remove_from_assignees(*remove)
+    if needed:
+        pr.add_to_assignees(*needed)
+
+for issue in github.search_issues(f"org:enarx is:pr is:public is:open -is:draft"):
+    pr = issue.as_pull_request()
+
+    # Get all reviewers that have not left a review on the current revision of
+    # the PR.
+    requests = get_requested_reviewers(pr)
+    
+    # Construct the set of responsible people for this PR.
+    responsible = requests | {RESPONSIBILITIES[r.state](pr.user.login, r.reviewer.login, requests) for (k, r) in get_latest_reviews(pr)}
+    responsible.remove(None)
```

Comment:
After this refactoring, my dictionary looks clunky. Let's break it apart with a loop and conditionals. We probably want this to be a separate function: `get_responsible_party(author, review, requests)` or something like that. This will also allow us to skip reviews in the 'APPROVED' state.